core/src/main/java/com/linecorp/armeria/client/endpoint/SlowStartAwareEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,175 @@
+        private final EndpointGroup endpointGroup;
+        private final AtomicInteger sequence = new AtomicInteger();
+        private volatile EndpointsAndWeights endpointsAndWeights;
+        private final EndpointWeightTransition endpointWeightTransition;
+
+        private volatile ScheduledFuture<?> updateEndpointWeightFuture;
+        private final Map<Endpoint, EndpointAndStep> currentWeights = new ConcurrentHashMap<>();
+
+        SlowStartAwareWeightedRoundRobinSelector(EndpointGroup endpointGroup,
+                                                 EndpointWeightTransition endpointWeightTransition) {
+            this.endpointGroup = endpointGroup;
+            this.endpointWeightTransition = endpointWeightTransition;
+            final List<Endpoint> endpoints = endpointGroup.endpoints();
+            endpointsAndWeights = new EndpointsAndWeights(endpoints);
+            endpointGroup.addListener(this::scheduleUpdateRequestWeight);
+        }
+
+        @Override
+        public EndpointGroup group() {
+            return endpointGroup;
+        }
+
+        @Override
+        public EndpointSelectionStrategy strategy() {
+            return SlowStartAwareEndpointSelectionStrategy.this;
+        }
+
+        @Override
+        public Endpoint select(ClientRequestContext ctx) {
+            final int currentSequence = sequence.getAndIncrement();
+            return endpointsAndWeights.selectEndpoint(currentSequence);
+        }
+
+        private void scheduleUpdateRequestWeight(List<Endpoint> newEndpoints) {
+            if (updateEndpointWeightFuture != null && !updateEndpointWeightFuture.cancel(false)) {
+                updateEndpointWeightFuture = null;
+            }
+            newEndpoints.forEach(e -> currentWeights.put(e, new EndpointAndStep(e)));
+            updateEndpointWeightFuture =
+                    executorService.scheduleAtFixedRate(() -> updateEndpointWeight(newEndpoints),
+                                                        0,
+                                                        slowStartInterval.toMillis(),
+                                                        TimeUnit.MILLISECONDS);
+        }
+
+        private void updateEndpointWeight(List<Endpoint> endpoints) {
+            endpointsAndWeights = new EndpointsAndWeights(
+                    endpoints.stream()
+                             .map(this::updateWeight)
+                             .collect(toImmutableList()));
+        }
+
+        private Endpoint updateWeight(Endpoint e) {
+            final EndpointAndStep endpointAndStep = currentWeights.get(e);
+            if (endpointAndStep == null) {
+                return e;
+            }
+            final int originalWeight = e.weight();
+            final int newWeight = endpointWeightTransition.compute(e, endpointAndStep.stepUp(), numberOfSteps);
+            if (newWeight >= originalWeight) {
+                currentWeights.remove(e);
```

Comment:
Based on my above comment, I think we want to keep the state that an endpoint has reached the defined weight and not reset it unless it goes away completely. I guess `EndpointAndStep` can have a flag marking that and instead of checking `null` above check the flag.